Crafting intercept grid

Tig709 opened the pull request:
## Related Issues

- #117 

## Summary of Changes

- Made an script which calculates interception point between 2 lines
- Made a grid with 2 lines
- Made scripts for getting a formula from a line and setting this formula text on the line
- Made an UI overlay for inserting answers and made code for checking the answer with the right answer
- Made events for right answer and wrong answer

## Additional Remarks

- FunctionLineController B currently is not working so I had to use transform.position.y as B value for lines in the minigames.

## Pull Criteria

- [x] Code commits have been reviewed
- [ ] Changes have been made if requested

MisakiCopperrose commented on Assets/Scripts/Crafter/LineToText.cs:
```diff
@@ -0,0 +1,36 @@
+
+        private TextMeshPro _formulaText;
+
+        [SerializeField] private int indexOfPointToFollow = 2;
```
Please put it with the other `SerializeField`

MisakiCopperrose reviewed (changes requested):
Please take a good look at it! 

If you have any questions I am happy to explain and help :)

Tig709 commented:
- Changed gameobject types to functionlinecontroller
- Made vector2exstenstion class for calculating interception between 2 lines
- Fixed hardcoded reference over editor assigned parent

MisakiCopperrose commented on Assets/Scripts/Crafter/LineToText.cs:
```diff
@@ -0,0 +1,35 @@
+        private float _a;
+        private float _b;
+        private int _indexOfPoint = 2;
+        private float _zPosition = -0.5f;
```
I think you misread which line I meant with this XD

But since I see you changed the `indexOfPointToFollow` to a serialised class field, make sure to think about whether you want outside influences (code from outside this class & people in the editor) to change the variable or not!

If they shouldn't change the variable, since it is integral to the logic of your code, make sure to use this: [Constants (C# Programming Guide)](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/constants) :)

Tig709 replied:
I think it can be usefull because if you have a longer line or have more points in the line renderer you probably want to have the formula on a different point of the line. If it is in the editor you can immediately do it in the editor while changing the line.

MisakiCopperrose replied:
Okidoki

MisakiCopperrose commented on Assets/Scripts/StringManipulation.cs:
```diff
@@ -0,0 +1,14 @@
+    /// </summary>
+    /// <param name="input">Input string</param>
+    /// <returns></returns>
+    public static string RemoveRegex(string input)
```
Please make sure the naming explains what the method does, in this case it would be something like `CleanUpDecimalOnlyString(...)`

MisakiCopperrose commented on Assets/Scripts/Crafter/LineToText.cs:
```diff
@@ -0,0 +1,36 @@
+            _line = lineToFollow.transform;
+            _formulaText = GetComponent<TextMeshPro>();
+            _a = _line.GetComponent<FunctionLineController>().A;
+            _b = _line.transform.position.y;
```
At line 23 you already called & cached the `transform` you are using here, so you are calling it twice!

Tig709 commented:
- Removed the GetComponent<FunctionLineController>'s from InterecptionCalculator
- Renamed regex function in stringmanipulation
- Moved crafterscene to right folder
- Fixed Hardcoded value not set to a constant or serialised editor field in LineToText
- Removed double transform call
